docs/design/2022-07-19-tiflash-support-fast-mode.md

```diff
@@ -0,0 +1,80 @@
+
+The capabilities of Fast Mode will be divided into multiple phases for continuous iteration and optimization.
+The query results of table scan in the first phase of Fast Mode are as follows:
+
+1. Does not support filtering of multiple versions of data
+2. Filter the data of the row where the filter delete operation is located (the del_mark in the row equals 1)
+
+## Product Behavior
+
+> Possible to change in this part
+
+We will add corresponding DDL statements to support TiFlash mode switching.
+
+```sql
+ALTER TABLE table_name SET TIFLASH MODE [mode]
+-- mode could be either FAST or NORMAL
+
+Samples:
+ALTER TABLE test.t SET TIFLASH MODE FAST;
+ALTER TABLE test.t SET TIFLASH MODE NORMAL;
+```
+
+Currently TiFlash does not support temporary tables, system tables, memory tables, and tables with non-utf-8 characters in their column names, thus executing this statement on these tables will return an error.
+
+Besides, we will add a field 'table_mode' to the system table information_schema.tiflash_replica to support the user viewing the TiFlash mode of the current table.
+
+```sql
+select * from information_schema.tiflash_replica
+```
+
+## Detailed Design
+
+### DDL Implementation
+
+The schema of the TiFlash table will be saved as part of the schema information of each table. In the table_info of the table, we will add a TiFlashMode field to identify the corresponding mode. This field will follow the sync schema approach among TiDB/TiKV/TiFlash as other schema infos.
+
+```c++
+enum class TiFlashMode
+{
+    Normal,
+    Fast,
+};
+```
+
+In order to facilitate checking mode, we will add the TABLE_MODE field to the system table information_schema.tiflash_replica to store the schema of the table.
+
+### Read Data Flow
+
+The TableScan operator in Normal Mode includes the following steps:
+
+1. Read data from Delta + Stable: Establish three data streams to read data from the MemTableSet and PersistedFileSet of the Delta layer, as well as the Stable layer.
+2. Sort Merge : Merge the 3 data streams established in step 1, and return the data in the order of (handle, version).
+3. Handle Filter : Filter the data according to the data range.
+4. MVCC + Column Filter: filter the data by MVCC according to version and del_mark, and filter out the unnecessary columns to return the data.
+
+Considering that MVCC filtering is a time-consuming step, but the impact on statistical results is limited, in Fast Mode, we propose to omit the step of MVCC filtering to speed up read operations. At the same time, because we omit the MVCC step, we also do not need to merge the data of the delta layer and the stable layer and return them in the order of (handle, version).
```

Comment:
```suggestion
Considering that MVCC filtering is a time-consuming step, but the impact on statistical results is limited, in Fast Mode, we propose omitting the MVCC filtering step to speed up read operations. At the same time, because we omit the MVCC step, we also do not need to merge the data of the delta layer and the stable layer and return them in the order of (handle, version).
Note that the rows returned under Fast Mode are not in order, TiDB should not generate plans with "keep order" TableScan operator when accessing those TiFlash tables.
```